### ISSUE TITLE
Include A Terminal Emulator

### ISSUE THREAD

**Issue by Deerp1976** [MuntashirAkon/AppManager#23]
Including a Terminal Emulator would be very useful even if you haven't got root.

My Huawei phone will not even allow me to check the build.prop from within the system folder as they have encrypted it but I can view it by running the script for it. 

This also give you the possibility for logcat viewing, netstat etc.

**Comment by 5c0tt-b0t** [comment]
I don't really agree with this, it should be purely **App Manager** based, terminals are for so much more than viewing file names in certain directories. Every phone has a 'shell' think of this as a built-in terminal that you don't see, when an app does something, it talks to the shell, hence this app doesn't need a terminal to cat/ls files, it already has that capability without a terminal. You asked already to use different file managers already, that will solve your problem, unless you root like me / turn off selinux.

**Comment by Deerp1976** [comment]
> I don't really agree with this, it should be purely **App Manager** based, terminals are for so much more than viewing file names in certain directories.
> Every phone has a 'shell' think of this as a built-in terminal that you don't see, when an app does something, it talks to the shell, hence this app doesn't need a terminal to cat/ls files, it already has that capability without a terminal.
> You asked to use different file managers already, that will solve your problem, unless you root like me / ADB root to turn off selinux.

I wish I could root my phone but Huawei had stopped giving out bootloader unlock codes years ago and as yet I have found no way to root. 

Your points are valid I understand them. It is an application Manager not a systems manager. I suppose I was just being greedy in wanting everything fitted into 1 app!

This app has already taught me quite a bit. Will be good to see how it progresses.

**Comment by 5c0tt-b0t** [comment]
No it is a valid point you make, as a terminal can achieve stuff **related** to app management, but what I mean is an app doesn't need a terminal built-in to view the stuff you can in a terminal, as it can ask the shell that is built-in to your device for these things. E.g, listing apps, files, getting stats etc.

**Comment by Deerp1976** [comment]
OK I think I understand you now! Technically I am very much an amateur. A very willing to learn but still very much just learning the basics!

Yes FX File Explorer has a script executor built in so understand you don't need a Terminal Emulator to run scripts etc. It's amazing what you can find out about the apps you have installed. This app informed me that a system app I have called Game Assistant which is a Huawei exclusive app has files and odex which is not suitable for my phone.

**Comment by 5c0tt-b0t** [comment]
> OK I think I understand you now! Technically I am very much an amateur. A very willing to learn but still very much just learning the basics!

No problem, we all start somewhere, compared to @MuntashirAkon ... I'm a big noob! we are all constantly learning!!

> Yes FX File Explorer has a script executor built in so understand you don't need a Terminal Emulator to run scripts etc.

Exactly! so what FX file manager is doing is running that script on the shell (mksh) that is built-in to your device, no terminal needed.

**Comment by MuntashirAkon** [comment]
Terminal emulators like Termux are beyond the scope of this project, I think, mostly because I'm trying to make the size of the app as small as possible. But a simple terminal emulator (like the one TWRP has) doesn't take much space, and I'm already using [Android Shell](https://github.com/jaredrummler/AndroidShell) library. So, it should not increase the app size that much. I'll add an **wontfix** label for now but I may consider it in future.

**Comment by UltraBlackLinux** [comment]
Would It be possible to just open the non disabled terminal emulator, like many music apps open the equalizer? I really need termux to run...

**Comment by MuntashirAkon** [comment]
> Would It be possible to just open the non disabled terminal emulator, like many music apps open the equalizer?

Yes, it's not difficult to add an option more Termux (if installed) in the main menu. But why would you want to run Termux from App Manager instead of your launcher? What are the advantages of such feature?

**Comment by UltraBlackLinux** [comment]
Why do Many Music players open the Installed equalizer? Cause they are to lazy to make one by temselves or want guarantee a working one. This is also the case here: We need a terminal emulator, but since many ppl like termux and you dont wanna add it to app manager cause its to big, just add a button that opens termux.

**Comment by Atrate** [comment]
Suggestion: to make the terminal opening actually useful, maybe include a function like "open termux session as app" (su - u0_a200 for example)

**Comment by MuntashirAkon** [comment]
> Suggestion: to make the terminal opening actually useful, maybe include a function like "open termux session as app" (su - u0_a200 for example)

This will go in the App Info tab, I believe.

**Comment by MuntashirAkon** [comment]
There are two options:

1. From what I understand from Termux' wiki [Intents and Hooks](https://wiki.termux.com/wiki/Intents_and_Hooks), I need to create an script somewhere in the shared file system (or within Termux' `$HOME`) and symlink it to `$HOME/bin/termux-url-opener` (create `bin` folder if not exists). This script should contain the following texts:

  ```sh
  #!/bin/sh
  file=$1
  exec su - u0_a$(cat ${file})
  ```
 In which case, I'll need to `Intent.ACTION_SEND` the `${file}` as `Intent.EXTRA_TEXT` (or any other methods). This is a hacky and root-only method but it's guaranteed to work.

2. Using `com.termux.permission.RUN_COMMAND` as described [here](https://github.com/termux/termux-app/pull/1029). But the problem with this method is that the user have to set `allow-external-apps` property to `true` in Termux config.

**Comment by MuntashirAkon** [comment]
It seems Termux' has several issues with `com.termux.permission.RUN_COMMAND`, especially the current release (v0.95) is mostly broken since `RUN_COMMAND_ARGUMENTS` doesn't actually work (see termux/termux-app/pull/1679). I'm posting related code below for future reference:

```java
        Intent intent = new Intent();
        intent.setClassName("com.termux", "com.termux.app.RunCommandService");
        intent.setAction("com.termux.RUN_COMMAND");
        intent.putExtra("com.termux.RUN_COMMAND_PATH", "/data/data/com.termux/files/usr/bin/bash");
        intent.putExtra("com.termux.RUN_COMMAND_ARGUMENTS", new String[]{"su", "-", "u0_a110"});  // Replace u0_a110 with actual user assigned to a package
        intent.putExtra("com.termux.RUN_COMMAND_BACKGROUND", false);
        try {
            startService(intent);
        } catch (Exception e) {
            e.printStackTrace();
        }
```

This requires Termux to be run in the background otherwise it will fail to open a session. It also requires `allow-external-apps=true` in `~/.termux/termux.properties`. In my end, the permission `com.termux.permission.RUN_COMMAND` has to be declared and requested like any dangerous permission.

**Comment by MuntashirAkon** [comment]
I also propose `run-as` option for debuggable apps which doesn't require root.

**Comment by MuntashirAkon** [comment]
It looks like a simple terminal emulator does have some use case (like the one in ADBungFu), especially for ADB users. The terminal emulator will be very simple with no completion support.

**Comment by NSTAdventure** [comment]
> It looks like a simple terminal emulator does have some use case (like the one in ADBungFu), especially for ADB users. The terminal emulator will be very simple with no completion support.

YEAH I WANT THIS FEATURE INSTEAD OF I PAY THAT GOD DAMN LADB APP ON PLAYSTORE 😤

**Comment by MuntashirAkon** [comment]
> YEAH I WANT THIS FEATURE INSTEAD OF I PAY THAT GOD DAMN LADB APP ON PLAYSTORE 😤

Please don't be so rude to a developer. Developers are human beings, therefore, they need money to live a life, and opening a play store account costs $25 which is a lot of money in most countries. That being said, [LADB](https://github.com/tytydraco/LADB) is a copy-lefted, free and open-source software just like App Manager. You can build the app from source or ask the developer to publish it on F-Droid.

**Comment by NSTAdventure** [comment]
> > YEAH I WANT THIS FEATURE INSTEAD OF I PAY THAT GOD DAMN LADB APP ON PLAYSTORE 😤
> 
> Please don't be so rude to a developer. Developers are human beings, therefore, they need money to live a life, and opening a play store account costs $25 which is a lot of money in most countries. That being said, [LADB](https://github.com/tytydraco/LADB) is a copy-lefted, free and open-source software just like App Manager. You can build the app from source or ask the developer to publish it on F-Droid.

Yeah i mean i don't want to pay i don't have enough money 

So if i can't i should searching alternative 


First i tried to use brevent apps

BUT idk
Cmd feature are paid or not

And 2nd i tried using terminal emulator but is looks like not working :(

**Comment by MuntashirAkon** [comment]
> Yeah i mean i don't want to pay i don't have enough money

As I said, LADB is a FOSS app if you can't buy it or build the app yourself, just write an email to the developer describing your financial condition. I'm sure he'll help.

**Comment by NSTAdventure** [comment]
> > Yeah i mean i don't want to pay i don't have enough money
> 
> As I said, LADB is a FOSS app if you can't buy it or build the app yourself, just write an email to the developer describing your financial condition. I'm sure he'll help.

for everyone sorry if i too rude it's not my intention at all 

i only want to say if i cannot purchase i won't crack a tools or any product is a sin to me :(


welp i want to send mail soon :( mommy

**Comment by shuvashish76** [comment]
> Terminal emulators like Termux are beyond the scope of this project, I think, mostly because I'm trying to make the size of the app as small as possible. But a simple terminal emulator (like the one TWRP has) doesn't take much space, and I'm already using [Android Shell](https://github.com/jaredrummler/AndroidShell) library. So, it should not increase the app size that much. I'll add an wontfix label for now but I may consider it in future.

This feature is already available in AM for a while.
candidate to /close the issue.

**Comment by MuntashirAkon** [comment]
> > Terminal emulators like Termux are beyond the scope of this project, I think, mostly because I'm trying to make the size of the app as small as possible. But a simple terminal emulator (like the one TWRP has) doesn't take much space, and I'm already using [Android Shell](https://github.com/jaredrummler/AndroidShell) library. So, it should not increase the app size that much. I'll add an wontfix label for now but I may consider it in future.
> 
> This feature is already available in AM for a while. candidate to /close the issue.

It's not an actual Terminal Emulator.
